### PR TITLE
Update semi-structured sparse tutorial

### PR DESCRIPTION
**Pull request by jcaip**
## Description
- Adds in .py script for semi-structured sparse
- new loss curves for dense / sparse training
- torch.compile integration and metrics

### CODE REVIEW

File: advanced_source/semi_structured_sparse.py [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+sparsity**.
+
+Semi-structured sparsity derives its name from its unique sparsity
+pattern, where n out of every 2n elements are pruned. We most often see
+n=2, hence 2:4 sparsity Semi-structured sparsity is particularly
+interesting because it can be efficiently accelerated on GPUs and
+doesn’t degrade model accuracy as much as other sparsity patterns.
+
+With the introduction of
+``semi-structured sparsity support <https://pytorch.org/docs/2.1/sparse.html#sparse-semi-structured-tensors>``\ \_,
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:20]:
```suggestion
`semi-structured sparsity support <https://pytorch.org/docs/2.1/sparse.html#sparse-semi-structured-tensors>`_,
```

File: advanced_source/semi_structured_sparse.py [hunk 2]
```diff
@@ -0,0 +1,618 @@
+"""
+(prototype) Accelerating BERT with semi-structured (2:4) sparsity
+=================================================================
+
+**Author**: ``Jesse Cai <https://github.com/jcaip>``\ \_
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:5]:
```suggestion
**Author**: `Jesse Cai <https://github.com/jcaip>`_
```

File: advanced_source/semi_structured_sparse.py [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+accelerate this 2:4 sparse model for inference, yielding a 1.3x speedup.
+
+Requirements
+------------
+
+-  PyTorch >= 2.1.
+-  A NVIDIA GPU with semi-structured sparsity support (Compute
+   Capability 8.0+).
+
+This tutorial is designed for beginners to semi-structured sparsity /
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:38]:
```suggestion
This tutorial is designed for beginners to semi-structured sparsity and
```

File: advanced_source/semi_structured_sparse.py [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+sparsity in general. For users with existing 2:4 sparse models,
+accelerating ``nn.Linear`` layers for inference with
+``to_sparse_semi_structured`` is as easy as:
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:41]:
```suggestion
``to_sparse_semi_structured`` is quite straightforward. Here is an example: 
```

File: advanced_source/semi_structured_sparse.py [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+# the pruned elements and store the specified elements in a compressed
+# format.
+# 
+# For semi-structured sparsity, we store exactly half of the original
+# parameters along with some compressed metadata about how the elements
+# were arranged.
+# 
+# .. image::
+# https://developer-blogs.nvidia.com/wp-content/uploads/2023/06/2-4-structured-sparsity-pattern.png
+# :align: center :width: 80%
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:103]:
```suggestion
# .. image:: https://developer-blogs.nvidia.com/wp-content/uploads/2023/06/2-4-structured-sparsity-pattern.png
#    :align: center :width: 80%
```

File: advanced_source/semi_structured_sparse.py [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+# 
+# ::
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:105]: This should be just a caption for the image as far as I understand. Caption syntax can be added after a blank line with 4 indents. In other words, removing line 104 and 105 should work.
```suggestion
```

File: advanced_source/semi_structured_sparse.py [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+# 
+#    Image sourced from `NVIDIA blog post <https://developer.nvidia.com/blog/structured-sparsity-in-the-nvidia-ampere-architecture-and-applications-in-search-engines/>`_ on semi-structured sparsity.
+# 
+# There are many different sparse layouts, each with their own benefits
+# and drawbacks. The 2:4 semi-structured sparse layout is particularly
+# interesting for two reasons: 1. Unlike previous sparse formats,
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:111]:
```suggestion
# interesting for two reasons:
# 
# 1. Unlike previous sparse formats,
```

File: .jenkins/metadata.json [hunk 8]
```diff
@@ -45,6 +45,9 @@
   "intermediate_source/scaled_dot_product_attention_tutorial.py": {
     "needs": "linux.g5.4xlarge.nvidia.gpu"
   },
+  "advanced_source/semi_structured_sparse.py": {
+    "needs": "linux.g5.4xlarge.nvidia.gpu"
+  }
```
Comment (svekars) [inline .jenkins/metadata.json:50]:
```suggestion
  } ,
```

File: advanced_source/semi_structured_sparse.py [hunk 9]
```diff
@@ -0,0 +1,618 @@
...
+# semi-structured sparsity was designed to be efficiently accelerated on
+# GPUs. In 2020, NVIDIA introduced hardware support for semi-structured
+# sparsity with their Ampere architecture, and have also released fast
+# sparse kernels via
+# CUTLASS/``cuSPARSELt <https://docs.nvidia.com/cuda/cusparselt/index.html>``\ *.
+# 2. At the same time, semi-structured sparsity tends to have a milder
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:117]:
```suggestion
# * At the same time, semi-structured sparsity tends to have a milder
```

Comment (svekars) [inline advanced_source/semi_structured_sparse.py:117]:
```suggestion
#
# * At the same time, semi-structured sparsity tends to have a milder
```

File: advanced_source/semi_structured_sparse.py [hunk 10]
```diff
@@ -0,0 +1,618 @@
...
+#    Image sourced from `NVIDIA blog post <https://developer.nvidia.com/blog/structured-sparsity-in-the-nvidia-ampere-architecture-and-applications-in-search-engines/>`_ on semi-structured sparsity.
+# 
+# There are many different sparse layouts, each with their own benefits
+# and drawbacks. The 2:4 semi-structured sparse layout is particularly
+# interesting for two reasons: 1. Unlike previous sparse formats,
+# semi-structured sparsity was designed to be efficiently accelerated on
+# GPUs. In 2020, NVIDIA introduced hardware support for semi-structured
+# sparsity with their Ampere architecture, and have also released fast
+# sparse kernels via
+# CUTLASS/``cuSPARSELt <https://docs.nvidia.com/cuda/cusparselt/index.html>``\ *.
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:116]:
```suggestion
# CUTLASS `cuSPARSELt <https://docs.nvidia.com/cuda/cusparselt/index.html>`__.
```

File: advanced_source/semi_structured_sparse.py [hunk 11]
```diff
@@ -0,0 +1,618 @@
...
+# 2. At the same time, semi-structured sparsity tends to have a milder
+# impact on model accuracy compared to other sparse formats, especially
+# when accounting for more advanced pruning / fine-tuning methods. NVIDIA
+# has shown in their ``white paper <https://arxiv.org/abs/2104.08378>``*
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:120]:
```suggestion
# has shown in their `white paper <https://arxiv.org/abs/2104.08378>`_
```

File: advanced_source/semi_structured_sparse.py [hunk 12]
```diff
@@ -0,0 +1,618 @@
+"""
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:1]:
```suggestion
# -*- coding: utf-8 -*-
"""
```

File: advanced_source/semi_structured_sparse.py [hunk 13]
```diff
@@ -0,0 +1,641 @@
...
+# SQuAD.
+# 
+# Intro & Setup
+# -------------
+# 
+# Let’s start by importing all the packages we need.
+# 
+
+!pip install datasets transformers evaluate accelerate pandas
+
```
Comment (svekars) [inline advanced_source/semi_structured_sparse.py:190]:
```suggestion

# If you are running this in Google Colab, run:
# .. code-block: python
# 
#    !pip install datasets transformers evaluate accelerate pandas
#

```

File: .jenkins/metadata.json [hunk 14]
```diff
@@ -45,6 +45,9 @@
   "intermediate_source/scaled_dot_product_attention_tutorial.py": {
     "needs": "linux.g5.4xlarge.nvidia.gpu"
   },
+  "advanced_source/semi_structured_sparse.py": {
+    "needs": "linux.g5.4xlarge.nvidia.gpu"
```
Comment (svekars) [inline .jenkins/metadata.json:49]:
```suggestion
    "needs": "linux.16xlarge.nvidia.gpu"
```

**Review by cpuhrsch — APPROVED** [review]

**Review by svekars — APPROVED** [review]
